### PR TITLE
feat(fmt): Add support for comments

### PR DESCRIPTION
**Pull request by jpopesculian**
Opening a draft to get some feedback on comments. Essentially the way comments are handled is by splitting them into two categories, prefixes and postfixes and types line and block comments

```rust
/* this is a prefix block comment */
let var = 1; // this is a postfix line comment
```

and trying to appropriately insert them where necessary, where prefixes need to start on a new line and line comments can't have anything follow them. As comments aren't in the parse tree the challenge is to try to figure out when to insert them appropriately and how to deal with the whitespace that they may or may not create.

This is essentially done by the adding a Comments context to the formatter and popping off comments as chunks get written via the `write_chunk` function and sometimes manually where more control is needed.

What I start doing here as well is to use the IndentWriter more by allowing the writing of multiple lines at a time as chunks can now be multiline given they contain line or prefix comments, and foregoing some of the assumptions of item size / line count before

The next steps are probably

* [ ] Replace all `write!` functions where necessary with `write_chunk!`
* [ ] Replace all `write_items_separated` with `write_chunks_separated`
* [ ] Lots and lots of tests
* [ ] Cleanup / refactor to find more patterns like `write_chunks_separated` (this will most likely be important when starting to really implement expressions and handling nesting / operator grouping)

### CODE REVIEW

**Comment by gakonst** [comment]
Could you rebase/cherry-pick the last few commits on latest master?

**Review by mattsse — COMMENTED** [review]
just some preliminary comments,

Additional functional documents would make the review somewhat easier later :)

File: fmt/src/comments.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                if this_line.trim_start().is_empty() {
+                    // comment sits on a new line
+                    if let Some(last_line) = lines_before.next() {
+                        if last_line.trim_start().is_empty() {
+                            // line before is empty
+                            (CommentPosition::Prefix, true)
+                        } else {
+                            // line has something
+                            let next_code = src[comment.loc().end()..]
+                                .lines()
```
Comment (jpopesculian) [inline fmt/src/comments.rs:96]: trim_comments needs to be called before lines, due to multi-line comments

File: fmt/src/comments.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                        chars.next();
+                        state = CommentState::Line;
+                    }
+                    Some('*') => {
+                        chars.next();
+                        state = CommentState::Block;
+                    }
+                    _ => out.push(ch),
+                },
+                _ => out.push(ch),
```
[the remark this answers is not in the document]
Comment (jpopesculian) [reply fmt/src/comments.rs:28]: I think that should be handled by the CommentState::Block which essentially drops all characters until `*/` is found or EOF

**Comment by gakonst** [comment]
@rkrasiuk wanna take a stab at the above?

**Comment by jpopesculian** [comment]
closing for #1833

**Pull request closed without merging**